// jinaai/dist/src/clients/BestBannerClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BestBannerClient = void 0;
var tslib_1 = require("tslib");
var HTTPClient_1 = require("./HTTPClient");
var BestBannerClient = (function (_super) {
    tslib_1.__extends(BestBannerClient, _super);
    function BestBannerClient(params) {
        var headers = params.headers, options = params.options, _a = params.baseURL, baseURL = _a === void 0 ? 'https://api.bestbanner.jina.ai/v1' : _a;
        var defaultHeaders = {
            'Content-Type': 'application/json',
        };
        var mergedHeaders = tslib_1.__assign(tslib_1.__assign({}, defaultHeaders), headers);
        return _super.call(this, { baseURL: baseURL, headers: mergedHeaders, options: options || {} }) || this;
    }
    BestBannerClient.prototype.fromArray = function (input, options) {
        return {
            data: input.map(function (i) { return (tslib_1.__assign({ text: i }, options)); })
        };
    };
    BestBannerClient.prototype.fromString = function (input, options) {
        return {
            data: [tslib_1.__assign({ text: input }, options)]
        };
    };
    BestBannerClient.prototype.isOutput = function (obj) {
        return typeof obj === 'object' &&
            obj.result &&
            obj.result.every(function (x) { return x.banners; });
    };
    BestBannerClient.prototype.toSimplifiedOutout = function (output) {
        if (!output.result || output.result.every(function (x) { return x.banners && x.banners.length != 0; }) == false)
            throw 'Remote API Error, bad output: ' + JSON.stringify(output);
        return {
            results: output.result.map(function (r) { return ({
                output: r.banners.map(function (b) { return b.url; }),
            }); })
        };
    };
    BestBannerClient.prototype.imagine = function (data, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var rawOutput, simplifiedOutput;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.post('/generate', data)];
                    case 1:
                        rawOutput = _a.sent();
                        simplifiedOutput = this.toSimplifiedOutout(rawOutput);
                        if ((options === null || options === void 0 ? void 0 : options.raw) == true)
                            simplifiedOutput.raw = rawOutput;
                        return [2, simplifiedOutput];
                }
            });
        });
    };
    return BestBannerClient;
}(HTTPClient_1.HTTPClient));
exports.BestBannerClient = BestBannerClient;
